Feature: Income Breakdown Report

Sonaryr opened the pull request:
Added a Sankey Report (see #1578) that shows how your income is broken down per category
Example (Categories and payees blanked out since it is my real budget):

<img width="1415" alt="Screenshot 2019-05-28 at 14 56 34" src="https://user-images.githubusercontent.com/10436899/58480210-250a9100-815a-11e9-8ab9-f306da7569c8.png">

All your incoming transactions get linked to "Income", then "Income" get's linked to "Expense" just so you get an immediate view on the difference.

The "Expense" is then broken down to your master categories which are then broken down per sub category.

joshmadewell commented:
Hey @Sonaryr this is awesome! It's a pretty requested feature so I'm excited to see it coming to life. I checked out this branch on my machine and with my budget (many payees, many categories). The graph ends up scaled down quite a bit if I look at all dates. I'm wondering if there's anything we can do to get rid of the massive blue section between "Income" and "Expense? 

I'm willing to merge this in as-is and try to get some feedback on it but either way. The graph is just really hard to reason about right now in a large budget.

Another option is to see if Highcharts supports any sort of zooming in on a graph, that would help too. When I get home from work, I'll try to remember to take a screenshot so you can see what exactly I'm talking about.

Sonaryr commented:
@joshmadewell yeah I get you I have it as well, I added the Income to Expense section to make the difference clear, because otherwise if the expense is bigger than the income it looks off, but does take a massive amount of space indeed.

I'll see what I can improve on it.

Sonaryr commented:
@joshmadewell I've changed the sorting on the categories, which makes for a cleaner graph.
I have also added some options to it so can choose to only see the income or expense or both, that way it will be a easier on the eye as well.

I've looked in to zooming, but highchart does not allow it in a sankey diagram.

Mansarde commented:
Just wanted to say thank you, this is an awesome feature! 😁👌

CygnusArtratus commented:
Thank you - I hadn't thought of this report but when I tried it wondered where it had been all my life! It would be great if the mouseover of a node showed the % of the previous node the total figure it shows already represents:

eg. Budget -> Category: **$1000** | _50%_

alansaul commented:
Awesome feature I had actually been thinking about attempting to implement this myself this week!

I would agree with @CygnusArtratus it would be great if there was a percentage of the preceding category as well.

One thing isn't clear to me. I have a extra category called "NET GAIN", it is unclear to me where this comes from, so I think could require some clarification.
@Sonaryr can you explain what this is? It appears to arise when `totalExpense <= totalIncome`, but all of my budget is allocated to a category so I figured this would be zero?
My only thinking is that if you haven't yet 'spent' the money in a budgeted category, it would show up under the "NET GAIN" category. I may be mistaken but it looks like my long-term budgeted savings, which have not yet become an expense, show up under the appropriate budgeted category, not `NET GAIN`?

alansaul commented:
@Sonaryr I think my confusion arises from the way categories expenses are calculated when the income has come directly into the category rather than `income -> to-be-budgeted -> category`.

I sometimes log income from other sources that were specifically for a single purpose as:
`person_A -> savings`

rather than 
`person_A -> to_be_budgeted`
assign to category savings

Perhaps this is my mistake, but the expenses in this PR are calculated as abs(expense) which means that
`person_A -> savings`
is listed as an expense (-money), even if it was an income (+money); which seems somewhat odd.

I can change my process of how I receive income to:
`person_A -> to_be_budgeted`
assign to category savings

but I kind of feel like it is a bug showing the absolute values?

In addition, if this is indeed the issue, it would be great is NET GAIN was split to show the categories in which the NET GAIN has been budgeted towards!

Mansarde commented:
@alansaul 
I think in order to let YNAB know that a particular inflow of money should be considered an increase in net worth (as opposed to simply coming *back* to you) it must first go into TBB to let YNAB know this.

That reflects what really happens, i.e. *new* money you get increases your net worth and now has to be budgeted (TBB).
Any money you put it into a category (no matter where it's coming from) is considered a *future expense*.
If there is a direct backflow into a category (skipping TBB) then it just means that the future expense has been reduced.

You can see this in the "Income vs Expense" page as well.
Money flowing into categories (as opposed to TBB first) will not show up in the "Income" table. It still increases your "Net Income" (because expenses were reduced) but it'll show up in the "Expense" table instead.

alansaul commented:
I've just modified my past expenses to reflect this and it looks much more sensible now (though I'm still confused why the `abs(expense)` would be used? Maybe it makes no difference though).

I guess its more of a feature request but it would be nice is "NET GAIN" was divided into the categories they are currently budgeted for. Though I guess Sankeys are primarily looking at expenses, I think this would be a nice feature since its ultimately a budgeting app!

Great feature.

Sonaryr commented:
Hi all,
Thanks for the awesome feedback on the feature!
The way income and expense are calculated is derived from the "income vs expense" report, so it follow the same logic. So it does not reflect budgetted money, but income and expenses.
**NET GAIN** and **OVERDRAFT** are 2 entries I added to make sure the graph makes more sense because otherwise you could see a wrong picture in some cases (e.g if you have more expenses then income it wouldn't show on first sight on the graph).

So if you expenses are higher then the income you will see en extra income entry **OVERDRAFT**. If your income is higher then your expenses you will see an extra high level expense category **NET GAIN** this means you made more money than you spend so kudos!

I have added a checkbox to show or hide these extra entries, so you can play around with them and see how they react.

jdoggg420 commented:
This is great!  Unfortunately I have so many categories that they don't all fit on the screen.  An option to scroll up and down would be great!

ZoneMR commented:
I'd love to see a revisit of what constitutes income vs expense.

For example, I have a category which tracks "House expenses via [my partner]". In some months, this will be an expense (I send my partner a contribution), and in some it will be negative (I receive a contribution).

This workflow really works for me - and for my purposes it's generally quite cleanly represented across YNAB. There's a few other categories I have where this approach makes sense to me also - a category which typically has insurance payments, but occasionally incoming funds resulting from a claim, a category tracking tax with an occasional refund, etc...

With this model, the income breakdown is totally broken - regardless of whether money was received or sent it's shown as 'spending'. It would be great to adjust the logic to work better in the circumstances I've described, rather than me marking such received funds as 'income', which un-links them from a category which I believe gives a fuller picture.

Sonaryr commented:
Hi @ZoneMR in the upcoming release of the toolkit positive categories will no longer be shown as an expense (this was a bug, see #1622 ). However right now all positive categories will be added as 1 single income source as discussed in that topic. So it might not quite be how you like it at the moment. But do create an issue with how you would like to see it and I will try to incorporate it in the report.
